clients/src/main/java/org/apache/kafka/common/utils/Time.java

```diff
@@ -58,6 +59,14 @@ default long hiResClockMs() {
      */
     void sleep(long ms);
 
+    /**
+     * Wait for a condition using the monitor of a given object. This avoids the implicit
+     * dependence on system time when calling {@link Object#wait()}.
+     *
+     * @throws org.apache.kafka.common.errors.TimeoutException if the timeout expires before the condition is satisfied
+     */
+    void waitObject(Object obj, Predicate<Void> condition, long timeoutMs) throws InterruptedException;
```

Comment:
Could we add description of the params to the javadoc? In particular the fact that we synchronize on `obj` and that we expect the caller to `notify` when condition is satisfied?